### PR TITLE
Allow parsers to capture data for future enrichment

### PR DESCRIPTION
**Pull request by buixor**
- Add a new `stash` section to parsers that can be used to capture data
 - Add new methods `GetFromCache` (and `SetInCache`) that allow to retrieve/set data from/to cache
 - The [cache](https://pkg.go.dev/github.com/bluele/gcache) itself supports LFU/LRU and TTL

A parser using the new feature would look like this :
```yaml
onsuccess: next_stage
debug: true
filter: "evt.Parsed.program == 'test'"
name: crowdsecurity/test-logs
description: "Parse test logs"
pattern_syntax:
  TEST_START: start %{DATA:program} thing with pid %{NUMBER:pid}
  TEST_CONTINUED: pid %{NUMBER:pid} did a forbidden thing
nodes:
  - grok:
      name: "TEST_START"
      apply_on: message
      statics:
        - meta: log_type
          value: test_start
    stash:
      - name: test_program_pid_assoc
        key: evt.Parsed.pid
        value: evt.Parsed.program
        ttl: 30s
        size: 10
  - grok:
      name: "TEST_CONTINUED"
      apply_on: message
      statics:
        - meta: log_type
          value: test_continue
        - meta: associated_prog_name
          expression: GetFromStash("test_program_pid_assoc", evt.Parsed.pid)
```
with an example log of :
```
start foobar thing with pid 12
start toto thing with pid 42
pid 12 did a forbidden thing
pid 42 did a forbidden thing
```
It would lead the 3rd and 4th line to be enriched with a `evt.Meta.associated_prog_name` field that would be set to `foobar` and `toto`.

TBD :
 - [x] Unit tests for the new cache package
 - [x] Functional tests for the capture / hydrate parsers 
 - [x] Make it use the associated parser's / scenario's logger so we can provide better contextual debug 
 - [ ] Should we be able to survive a reload -> not worth for now

### CODE REVIEW

File: pkg/parser/node.go [hunk 1]
```diff
@@ -103,6 +107,25 @@ func (n *Node) validate(pctx *UnixParserCtx, ectx EnricherCtx) error {
...
+
+	for idx, stash := range n.Stash {
+		if stash.Name == "" {
+			return fmt.Errorf("stash %d : name must be set", idx)
+		}
+		if stash.Value == "" {
+			return fmt.Errorf("stash %s : expression must be set", stash.Name)
+		}
+		if stash.Key == "" {
+			return fmt.Errorf("stash %s : expression must be set", stash.Name)
```
Comment (blotus) [inline pkg/parser/node.go:119]: Key instead of expression

File: pkg/parser/node.go [hunk 2]
```diff
@@ -103,6 +107,25 @@ func (n *Node) validate(pctx *UnixParserCtx, ectx EnricherCtx) error {
 			}
 		}
 	}
+
+	for idx, stash := range n.Stash {
+		if stash.Name == "" {
+			return fmt.Errorf("stash %d : name must be set", idx)
+		}
+		if stash.Value == "" {
+			return fmt.Errorf("stash %s : expression must be set", stash.Name)
```
Comment (blotus) [inline pkg/parser/node.go:116]: Value instead of expression

File: pkg/parser/node.go [hunk 3]
```diff
@@ -285,6 +308,46 @@ func (n *Node) process(p *types.Event, ctx UnixParserCtx, expressionEnv map[stri
 		clog.Tracef("! No grok pattern : %p", n.Grok.RunTimeRegexp)
 	}
 
+	//Process the stash (data collection) if present
+	for idx, stash := range n.Stash {
+		var value string
+		var key string
+		if stash.ValueExpression == nil {
```
Comment (blotus) [inline pkg/parser/node.go:315]: Is it necessary as value must be set ?

File: pkg/parser/node.go [hunk 4]
```diff
@@ -285,6 +308,46 @@ func (n *Node) process(p *types.Event, ctx UnixParserCtx, expressionEnv map[stri
...
+			clog.Warningf("Stash %d has no value expression, skipping", idx)
+			continue
+		}
+		if stash.KeyExpression == nil {
```
Comment (blotus) [inline pkg/parser/node.go:319]: Is it necessary as key must be set ?

File: pkg/parser/node.go [hunk 5]
```diff
@@ -285,6 +308,46 @@ func (n *Node) process(p *types.Event, ctx UnixParserCtx, expressionEnv map[stri
...
+		output, err := expr.Run(stash.ValueExpression, cachedExprEnv)
+		if err != nil {
+			clog.Warningf("Error while running stash val expression : %v", err)
+		}
+		//can we expect anything else than a string ?
+		switch output.(type) {
+		case string:
+			value = output.(string)
+		default:
+			clog.Warningf("unexpected type %t (%v) while running '%s'", output, output, stash.ValueExpression)
```
Comment (blotus) [inline pkg/parser/node.go:333]: Missing continue ?

File: pkg/parser/node.go [hunk 6]
```diff
@@ -285,6 +308,46 @@ func (n *Node) process(p *types.Event, ctx UnixParserCtx, expressionEnv map[stri
...
+		output, err = expr.Run(stash.KeyExpression, cachedExprEnv)
+		if err != nil {
+			clog.Warningf("Error while running stash key expression : %v", err)
+		}
+		//can we expect anything else than a string ?
+		switch output.(type) {
+		case string:
+			key = output.(string)
+		default:
+			clog.Warningf("unexpected type %t (%v) while running '%s'", output, output, stash.KeyExpression)
```
Comment (blotus) [inline pkg/parser/node.go:346]: Missing continue ?

File: pkg/exprhelpers/exprlib.go [hunk 7]
```diff
@@ -69,6 +70,8 @@ func GetExprEnv(ctx map[string]interface{}) map[string]interface{} {
 		"GetDecisionsSinceCount": GetDecisionsSinceCount,
 		"Sprintf":                fmt.Sprintf,
 		"ParseUnix":              ParseUnix,
+		"GetFromCache":           cache.GetKey,
+		"SetInCache":             cache.SetKey,
```
Comment (LaurenceJJones) [inline pkg/exprhelpers/exprlib.go:74]: is the term stash or cache?

File: pkg/parser/node.go [hunk 8]
```diff
@@ -103,6 +107,25 @@ func (n *Node) validate(pctx *UnixParserCtx, ectx EnricherCtx) error {
 			}
 		}
 	}
+
+	for idx, stash := range n.Stash {
```
Comment (blotus) [inline pkg/parser/node.go:111]: We probably want to only process the stash if the grok pattern succeeded ?

File: pkg/exprhelpers/exprlib.go (hunk 7 above)
Comment (buixor) [reply pkg/exprhelpers/exprlib.go:74]: stop confronting me with my inconsistencies :sweat_smile:

File: pkg/parser/node.go (hunk 8 above)
Comment (buixor) [reply pkg/parser/node.go:111]: Asked the same question, but we might want to capture some date based on a grok that is done at a higher level :


```yaml
onsuccess: next_stage
debug: true
filter: "evt.Parsed.program == 'test'"
name: crowdsecurity/test-logs
description: "Parse test logs"
grok:
  pattern: something something capture %{DATA:event _type}
nodes:
  -  filter: evt.Parsed.event_type == "case1"
     static:
      ...
  - filter: evt.Parsed.event_type == "case2"
    stash:
      - name: ...

```

**Review by blotus — APPROVED** [review]

**Pull request merged**